### PR TITLE
Bugfix: restore ability to run virtual tests

### LINKED ISSUE #35519: spack test: Spec.package can only be called on concrete specs
### Steps to reproduce

Unclear exactly which packages are required to reproduce this. I have a massive environment I'm testing:
```yaml
spack:
  specs:
  - gdal
  - opencv
  - py-black
  - py-cartopy
  - py-cmocean
  - py-einops
  - py-fiona
  - py-flake8
  - py-geocube
  - py-geopandas
  - py-h5py
  - py-ipywidgets
  - py-isort
  - py-jax
  - py-kornia
  - py-laspy
  - py-matplotlib
  - py-mypy
  - py-nbmake
  - py-nbsphinx
  - py-netcdf4
  - py-numpy
  - py-omegaconf
  - py-pandas
  - py-pillow-simd
  - py-pycocotools
  - py-pydocstyle
  - py-pygeos
  - py-pyproj
  - py-pyro-ppl
  - py-pytest
  - py-pytest-cov
  - py-pytorch-lightning
  - py-pytorch-sphinx-theme
  - py-pyupgrade
  - py-radiant-mlhub
  - py-rarfile
  - py-rasterio
  - py-rtree
  - py-setuptools
  - py-scikit-image
  - py-scikit-learn
  - py-scipy
  - py-segmentation-models-pytorch
  - py-shapely
  - py-sphinx
  - py-tensorboard
  - py-timm
  - py-torch
  - py-torchmetrics
  - py-torchvision
  - py-xarray
  - py-zipfile-deflate64
  view: true
  concretizer:
    unify: true
```
I run `spack install` and `spack test run`.

### Error message

```console
$ spack -d test run
...
==> [2023-02-16-11:55:18.673419] Installing /projects/dali/spack/var/spack/repos/builtin/packages/openjdk/package.py to /home/adamjs5/.spack/test/32ccsglunnecbl3oi2fztym6ti6alinu/repo/openjdk
==> [2023-02-16-11:55:18.677207] Installing /projects/dali/spack/var/spack/repos/builtin/packages/cuda/package.py to /home/adamjs5/.spack/test/32ccsglunnecbl3oi2fztym6ti6alinu/repo/cuda
Traceback (most recent call last):
  File "/projects/dali/spack/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/projects/dali/spack/lib/spack/spack_installable/main.py", line 37, in main
    sys.exit(spack.main.main(argv))
  File "/projects/dali/spack/lib/spack/spack/main.py", line 1004, in main
    return _main(argv)
  File "/projects/dali/spack/lib/spack/spack/main.py", line 959, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/projects/dali/spack/lib/spack/spack/main.py", line 987, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/projects/dali/spack/lib/spack/spack/main.py", line 636, in _invoke_command
    return_val = command(parser, args)
  File "/projects/dali/spack/lib/spack/spack/cmd/test.py", line 476, in test
    globals()["test_%s" % args.test_command](args)
  File "/projects/dali/spack/lib/spack/spack/cmd/test.py", line 226, in test_run
    externals=args.externals,
  File "/projects/dali/spack/lib/spack/spack/install_test.py", line 169, in __call__
    self.write_reproducibility_data()
  File "/projects/dali/spack/lib/spack/spack/install_test.py", line 393, in write_reproducibility_data
    spack.repo.path.dump_provenance(vspec, repo_cache_path)
  File "/projects/dali/spack/lib/spack/spack/repo.py", line 304, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/projects/dali/spack/lib/spack/spack/repo.py", line 859, in dump_provenance
    return self.repo_for_pkg(spec).dump_provenance(spec, path)
  File "/projects/dali/spack/lib/spack/spack/repo.py", line 304, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/projects/dali/spack/lib/spack/spack/repo.py", line 1068, in dump_provenance
    for patch in itertools.chain.from_iterable(spec.package.patches.values()):
  File "/projects/dali/spack/lib/spack/spack/spec.py", line 1615, in package
    assert self.concrete, "Spec.package can only be called on concrete specs"
AssertionError: Spec.package can only be called on concrete specs
```
I tried `spack clean -m` in case the cache contained an old patch that no longer existed but it didn't help.

### Information on your system

* **Spack:** 0.20.0.dev0
* **Python:** 3.6.8
* **Platform:** linux-rhel7-zen3
* **Concretizer:** clingo

@tldahlgren 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by tldahlgren**
Fixes #35519 (at least should)

This PR restores the ability to run stand-alone tests on virtual provider packages, including the ability to "inherit" tests from the virtual packages.

Also fixes an issue reported by @jfinney10 on slack:

```
$ spack -d test run
[...snip...]
  File "$spack/lib/spack/spack/repo.py", line 1153, in dump_provenance
    spec.package.patches.values()):
  File "$spack/lib/spack/spack/spec.py", line 1520, in package
    assert self.concrete, "Spec.package can only be called on concrete specs"
AssertionError: Spec.package can only be called on concrete specs
```

TODO

- [x] Stand-alone testing
- [x] Ensure pulled all of the relevant code from (local) #34236

### CODE REVIEW

**Comment by tldahlgren** [comment]
Not clear why the `libdwarf` source stage directory was missing for the `macOS` check  (https://github.com/spack/spack/actions/runs/4140755709/jobs/7159751165).

**Comment by tldahlgren** [comment]
@tgamblin @alalazo Looks like all of the checks are now passing.  This PR restores `spack test`'s running of stand-alone tests defined in virtual packages while removing errors related to the fact those specs aren't concretized.

**Comment by tldahlgren** [comment]
Grrr.  I'm seeing positive coverage for all of the affected files and don't seem to be able to force re-running the check.

File: lib/spack/spack/install_test.py [hunk 1]
```diff
@@ -110,6 +111,95 @@ def get_test_suite(name):
...
+    """Return the spec's package class.
+
+    Args:
+        spec (spack.spec.Spec): spec instance
+    """
+    try:
+        cls = spec.package.__class__
+    except AssertionError:
+        try:
+            cls = spec.package_class
```
Comment (alalazo) [inline lib/spack/spack/install_test.py]: Are there cases where these two might be different?

File: lib/spack/spack/install_test.py [hunk 2]
```diff
@@ -110,6 +111,95 @@ def get_test_suite(name):
...
+            test_specs = [pkg.spec] + [spack.spec.Spec(v_name) for v_name in sorted(v_names)]
+            for spec in test_specs:
+                methods.extend(test_functions(spec, names=names))
+            return methods
+
+    elif inspect.isclass(spec_or_pkg):
+        cls = spec_or_pkg
+
+    else:
+        raise ValueError("Cannot retrieve test methods for {0}".format(spec_or_pkg))
```
Comment (alalazo) [inline lib/spack/spack/install_test.py]: Is it worth extracting this part into its own function? If I understand correctly, this gets as input either a spec object or a package object or a package class and returns a package class.

File: lib/spack/spack/package_base.py [hunk 3]
```diff
@@ -114,6 +119,11 @@ def deprecated_version(pkg, version):
     return False
 
 
+def package_directory(cls):
+    """Returns the path to the package class directory."""
+    return os.path.abspath(os.path.dirname(cls.module.__file__))
```
Comment (alalazo) [inline lib/spack/spack/package_base.py]: Why was extracting this function necessary? I think `package_dir` would work on package classes too.

**Review by alalazo — COMMENTED** [review]
Just a few initial comments and questions. There's a conflict to be solved.

File: lib/spack/spack/install_test.py (hunk 1 above)
Comment (tldahlgren) [reply lib/spack/spack/install_test.py]: I don't think so.  I was thinking the majority of time the Spec is concrete and readily retrieved from `spec.package`.  

When it isn't, such as for virtual packages, then the call `spec.package_class` that makes multiple calls (e.g., `spack.repo.path.get_pkg_class()`, importing the module, and more) is performed.

File: lib/spack/spack/install_test.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/install_test.py]: Good point.

File: lib/spack/spack/package_base.py (hunk 3 above)
Comment (tldahlgren) [reply lib/spack/spack/package_base.py]: Can't recall.  Let me look into it.

File: lib/spack/spack/install_test.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/install_test.py]: That needed another pass over the logic anyway.

**Comment by tldahlgren** [comment]
I need to update my local `black` (or some setting) because it is not reporting the same style issues reported here.  There was a style fix reported yesterday that I made but it seems like a new layout was required today.

**Comment by adamjstewart** [comment]
black 23 is more strict than black 22, I recently updated CI to use black 23 and updated all existing code.

File: lib/spack/spack/package_base.py (hunk 3 above)
Comment (tldahlgren) [reply lib/spack/spack/package_base.py]: Removed

File: lib/spack/spack/install_test.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/install_test.py]: Extracted into its own function.

File: lib/spack/spack/install_test.py [hunk 4]
```diff
@@ -110,6 +111,105 @@ def get_test_suite(name):
...
 
+def package_class(spec_or_pkg):
+    """Return the package class for the spec or package.
+
+    Args:
+        spec_or_pkg (spack.spec.Spec or spack.package_base.PackageBase): spec or
+            package (class)
+
+    Raises: ValueError: If not given a spec, package, or class
+    """
```
Comment (alalazo) [inline lib/spack/spack/install_test.py]: We can add type annotation like this:
```diff
diff --git a/lib/spack/spack/install_test.py b/lib/spack/spack/install_test.py
index 416d502acc..ddbc717195 100644
--- a/lib/spack/spack/install_test.py
+++ b/lib/spack/spack/install_test.py
@@ -8,6 +8,7 @@
 import os
 import re
 import shutil
+from typing import Optional, Type, Union
 
 import llnl.util.filesystem as fs
 import llnl.util.tty as tty
@@ -111,14 +112,19 @@ def get_test_suite(name):
     return names[0]
 
 
-def package_class(spec_or_pkg):
+PkgClass = Type["spack.package_base.PackageBase"]
+
+
+def package_class(
+    spec_or_pkg: "Union[Spec, spack.package_base.PackageBase, PkgClass]",
+) -> Optional[PkgClass]:
     """Return the package class for the spec or package.
 
     Args:
-        spec_or_pkg (spack.spec.Spec or spack.package_base.PackageBase): spec or
-            package (class)
+        spec_or_pkg: spec or package (class)
 
-    Raises: ValueError: If not given a spec, package, or class
+    Raises:
+        ValueError: If not given a spec, package, or class
     """
     if isinstance(spec_or_pkg, spack.spec.Spec):
         spec = spec_or_pkg

```
so that `mypy` will check a bit more of our code base.

File: lib/spack/spack/install_test.py [hunk 5]
```diff
@@ -110,6 +111,105 @@ def get_test_suite(name):
...
+            try:
+                cls = spec.package_class
+            except spack.repo.UnknownPackageError as e:
+                tty.debug("{0}: Cannot retrieve class from spec: {1}".format(spec.name, str(e)))
+                return None
+
+    elif isinstance(spec_or_pkg, spack.package_base.PackageBase):
+        cls = spec_or_pkg.__class__
+
+    elif inspect.isclass(spec_or_pkg):
```
Comment (alalazo) [inline lib/spack/spack/install_test.py]: Would this:
```suggestion
    elif issubclass(spec_or_pkg, spack.package_base.PackageBase):
```
work too?

File: lib/spack/spack/install_test.py [hunk 6]
```diff
@@ -110,6 +111,105 @@ def get_test_suite(name):
...
+        try:
+            cls = spec.package.__class__
+        except AssertionError:
+            try:
+                cls = spec.package_class
+            except spack.repo.UnknownPackageError as e:
+                tty.debug("{0}: Cannot retrieve class from spec: {1}".format(spec.name, str(e)))
+                return None
+
+    elif isinstance(spec_or_pkg, spack.package_base.PackageBase):
```
Comment (alalazo) [inline lib/spack/spack/install_test.py:139]: I think we never import `spack.package_base` in this Python module (and importing it at the module level will probably lead to circular dependency issues).

File: lib/spack/spack/install_test.py [hunk 7]
```diff
@@ -110,6 +111,105 @@ def get_test_suite(name):
...
+    Raises: ValueError: If not given a spec, package, or class
+    """
+    if add_virtuals and isinstance(spec_or_pkg, spack.package_base.PackageBase):
+        pkg = spec_or_pkg
+        v_names = virtuals(pkg)
+        test_specs = [pkg.spec] + [spack.spec.Spec(v_name) for v_name in sorted(v_names)]
+        methods = []
+        for spec in test_specs:
+            methods.extend(test_functions(spec, add_virtuals=False, names=names))
+        return methods
```
Comment (alalazo) [inline lib/spack/spack/install_test.py:178]: (Minor) If I read this one correctly here we are expanding virtual packages and recursing with the `add_virtuals` argument flipped, so to fall through and hit the code below. Would it be easy to avoid recursion and separate:
1. Expansion of virtual specs
2. Extraction of tests functions from list of specs

? If it seems complicated, let's leave it for another PR.

File: lib/spack/spack/install_test.py [hunk 8]
```diff
@@ -110,6 +111,105 @@ def get_test_suite(name):
...
+
+    cls = package_class(spec_or_pkg)
+    if not cls:
```
Comment (alalazo) [inline lib/spack/spack/install_test.py]: (Minor) I think this might be slightly safer for future refactoring:
```suggestion
    if cls is None:
```

File: lib/spack/spack/install_test.py [hunk 9]
```diff
@@ -110,6 +111,105 @@ def get_test_suite(name):
...
+    methods = inspect.getmembers(cls, predicate=lambda x: inspect.isfunction(x))
+    tests = []
+    for name, test_fn in methods:
+        if not name == "test":
+            continue
+
+        source = (inspect.getsource(test_fn)).splitlines()[1:]
+        lines = (ln.strip() for ln in source)
+        statements = [ln for ln in lines if not ln.startswith("#")]
+        empty = len(statements) > 0 and statements[0] == "pass"
```
Comment (alalazo) [inline lib/spack/spack/install_test.py:199]: Curious why we need to inspect sources. Is it to avoid printing to screen? Or something else?

File: lib/spack/spack/repo.py [hunk 10]
```diff
@@ -1065,15 +1065,25 @@ def dump_provenance(self, spec, path):
...
+                if patch.path:
+                    if os.path.exists(patch.path):
+                        fs.install(patch.path, path)
+                    else:
+                        tty.warn("Patch file did not exist: %s" % patch.path)
+        except AssertionError as e:
+            # virtual specs won't have patches to install
+            tty.debug(
+                "Could not copy patch files for virtual package {0}: {1}".format(spec.name, str(e))
+            )
```
Comment (alalazo) [inline lib/spack/spack/repo.py]: Can we do this more explicitly with a conditional upfront? This is for 2 main reasons:
1. We might catch another assertion in the code
2. Assertion might be turned off from the Python interpreter

**Review by alalazo — CHANGES_REQUESTED** [review]
I have test driven the PR and was able to:
```console
$ spack -v test mpich
[ ... ]
$ spack test results --logs 4lb3fp6vigamkxjn56bxkdvvxtpwmkep
==> Results for test suite '4lb3fp6vigamkxjn56bxkdvvxtpwmkep':
==> test specs:
==>   mpich-4.1-yva65xh PASSED
==> Testing package mpich-4.1-yva65xh
==> [2023-02-27-18:11:01.638157] Installing /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/mpich-4.1-yva65xhfzl7xertquaahwt3rg5rwpr6q/.spack/test to /home/culpo/.spack/test/4lb3fp6vigamkxjn56bxkdvvxtpwmkep/mpich-4.1-yva65xh/cache/mpich
==> [2023-02-27-18:11:01.902862] test: generate finalized file
==> [2023-02-27-18:11:01.903060] '/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/mpich-4.1-yva65xhfzl7xertquaahwt3rg5rwpr6q/bin/mpicc' '/home/culpo/.spack/test/4lb3fp6vigamkxjn56bxkdvvxtpwmkep/mpich-4.1-yva65xh/cache/mpich/test/mpi/init/finalized.c' '-Wall' '-g' '-o' 'finalized'
PASSED
==> [2023-02-27-18:11:02.020309] test: run finalized example
==> [2023-02-27-18:11:02.020624] './finalized'
 No Errors
PASSED
==> [2023-02-27-18:11:03.382000] test: generate sendrecv file
==> [2023-02-27-18:11:03.382351] '/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/mpich-4.1-yva65xhfzl7xertquaahwt3rg5rwpr6q/bin/mpicc' '/home/culpo/.spack/test/4lb3fp6vigamkxjn56bxkdvvxtpwmkep/mpich-4.1-yva65xh/cache/mpich/test/mpi/basic/sendrecv.c' '-Wall' '-g' '-o' 'sendrecv'
PASSED
==> [2023-02-27-18:11:03.530278] test: run sendrecv example
==> [2023-02-27-18:11:03.530589] './sendrecv'
Simple Send/Recv test.
Two processes needed.
PASSED
==> [2023-02-27-18:11:04.027395] test: mpicc: expect command status in [0]
==> [2023-02-27-18:11:04.027744] '/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/mpich-4.1-yva65xhfzl7xertquaahwt3rg5rwpr6q/bin/mpicc' '-o' 'mpi_hello_c' '/home/culpo/.spack/test/4lb3fp6vigamkxjn56bxkdvvxtpwmkep/mpich-4.1-yva65xh/data/mpi/mpi_hello.c'
PASSED
==> [2023-02-27-18:11:04.154281] test: mpirun: expect command status in [0]
==> [2023-02-27-18:11:04.154603] '/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/mpich-4.1-yva65xhfzl7xertquaahwt3rg5rwpr6q/bin/mpirun' '-np' '1' 'mpi_hello_c'
Hello world! From rank 0 of 1
PASSED
==> [2023-02-27-18:11:04.647879] test: mpif90: expect command status in [0]
==> [2023-02-27-18:11:04.648273] '/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/mpich-4.1-yva65xhfzl7xertquaahwt3rg5rwpr6q/bin/mpif90' '-o' 'mpi_hello_f' '/home/culpo/.spack/test/4lb3fp6vigamkxjn56bxkdvvxtpwmkep/mpich-4.1-yva65xh/data/mpi/mpi_hello.f'
PASSED
==> [2023-02-27-18:11:04.769080] test: mpirun: expect command status in [0]
==> [2023-02-27-18:11:04.769346] '/home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/mpich-4.1-yva65xhfzl7xertquaahwt3rg5rwpr6q/bin/mpirun' '-np' '1' 'mpi_hello_f'
 Hello world! From rank           0 of            1
PASSED
==> [2023-02-27-18:11:05.232167] Completed testing

============================= 1 passed of 1 specs ==============================

```
I left a few minor comments on the code, but from a user perspective :heavy_check_mark:

File: lib/spack/spack/install_test.py (hunk 4 above)
Comment (tldahlgren) [reply lib/spack/spack/install_test.py]: I need to get into the habit of adding the annotations.

File: lib/spack/spack/install_test.py (hunk 9 above)
Comment (tldahlgren) [reply lib/spack/spack/install_test.py:199]: Ppl have expressed concerns about distinguishing between having a test method and having one that actually does something so inspecting to make sure non-empty (non-`pass`) helps make that distinction.

File: lib/spack/spack/repo.py (hunk 10 above)
Comment (tldahlgren) [reply lib/spack/spack/repo.py]: I was thinking I'd be more `pythonic` :wink: but I can certainly check in advance.

**Comment by tldahlgren** [comment]
Closing this PR since the changes originated in #34236, which has refactoring that causes redundant work iterating on this and the other PR.
